Tdk invensense icp101xx pressure sensor

rbuisson-invn opened the pull request:
drivers: sensor: icp101xx: supports icp101xx sensors
dts: supports icp101xx invn pressure sensor
module: HAL TDK module

Adds official TDK Invensense Inc. driver in TDK HAL module for icp101xx sensor.
Replacing icp10125 unofficial driver
Validated with custom setup: nrf52dk_nrf52832 + icp10125 Piromoni board

Signed-off-by: Remi Buisson [remi.buisson@tdk.com](mailto:remi.buisson@tdk.com)

rbuisson-invn commented on tests/drivers/build_all/sensor/i2c.dtsi:
Fixed, thanks for your review.

rbuisson-invn commented on drivers/sensor/tdk/icp101xx/icp101xx_drv.h:
```diff
@@ -0,0 +1,40 @@
+#include "Icp101xx.h"
+#include "Icp101xxSerif.h"
+
+#define DT_DRV_COMPAT invensense_icp101xx
```
Fixed, thanks for your review.

teburd commented on drivers/sensor/tdk/icp101xx/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+zephyr_library()
+
+if (CONFIG_CPU_HAS_FPU STREQUAL "y")
+  zephyr_compile_definitions(ICP101XX_DRV_USE_FLOATS)
```
We shouldn't be using floats at all in these drivers

rbuisson-invn replied:
Floats are mainly used to provide altitude.
I know floats are to be avoided to comply with platform not having efficient float support.
The trade-off here is to provide altitude only when an FPU is present, and that floats are efficiently supported.
If that solution is not acceptable for Zephyr, I can remove altitude support from the driver, even though that feature could be useful for users.

teburd replied:
I think what you have is reasonable after looking again as it is opt-in. I'd simply note that as you already understand, not every architecture zephyr supports has an fpu so we do strive to use integer math and avoid the code and time cost for software float.

teburd commented on modules/hal_tdk/Kconfig:
```diff
@@ -1,6 +1,7 @@
 # Copyright (c) 2024 TDK Invensense
 # SPDX-License-Identifier: Apache-2.0
 
+
```
stray new line add

rbuisson-invn replied:
Removed, thanks.

teburd commented on drivers/sensor/tdk/icp101xx/icp101xx_drv.h:
```diff
@@ -0,0 +1,38 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_DRIVERS_SENSOR_ICP101XXX_ICP101XXX_H_
```
There doesn't seem to be a good reason to split this into a .h file, unless there's multiple .c files needing this it should be in the same compliation unit to keep things simple

teburd commented on drivers/sensor/tdk/icp101xx/icp101xx_drv.c:
```diff
@@ -0,0 +1,261 @@
+	k_sleep(K_MSEC(get_conversion_ms(data->icp_device.measurement_mode)));
+	do {
+		k_sleep(K_USEC(200));
+		rc = inv_icp101xx_get_data(&data->icp_device, &(data->raw_pressure),
```
Is this call out to the tdk hal really needed? Isn't this simply reading a few registers?

rbuisson-invn replied:
This create a specific command sent over I2C.
It seems better to have a dedicated function.

teburd commented on drivers/sensor/tdk/icp101xx/icp101xx_drv.c:
```diff
@@ -0,0 +1,261 @@
+	      (chan == SENSOR_CHAN_ALTITUDE) || (chan == SENSOR_CHAN_ALL))) {
+		return -ENOTSUP;
+	}
+	rc = inv_icp101xx_enable_sensor(&data->icp_device, 1);
```
Isn't this equivalent to i2c_write() with a simple register write? Why is the TDK HAL needed here?

teburd reviewed (commented):
Please consider dropping float math, unclear what the tdk hal is providing here to be frank.

teburd reviewed (dismissed):
reasonable answers/fixes given I think to my review comments

rbuisson-invn commented:
> reasonable answers/fixes given I think to my review comments

Thanks for the review and approval !

teburd reviewed (dismissed):
Refresh +1

MaureenHelm commented on drivers/sensor/tdk/icp101xx/icp101xx_drv.c:
```diff
@@ -0,0 +1,259 @@
+	return 0;
+}
+
+static const struct sensor_driver_api icp101xx_api_funcs = {
```
```suggestion
static DEVICE_API(sensor, icp101xx_api_funcs) = {
```

MaureenHelm commented on drivers/sensor/tdk/icp101xx/icp101xx_drv.c:
```diff
@@ -0,0 +1,259 @@
+		.i2c = I2C_DT_SPEC_INST_GET(inst),                                                 \
+		.mode = DT_INST_ENUM_IDX(inst, mode),                                              \
+	};                                                                                         \
+	DEVICE_DT_INST_DEFINE(inst, icp101xx_init, NULL, &icp101xx_drv_##inst,                     \
```
```suggestion
	SENSOR_DEVICE_DT_INST_DEFINE(inst, icp101xx_init, NULL, &icp101xx_drv_##inst,                     \
```

MaureenHelm commented on dts/bindings/sensor/invensense,icp10125.yaml:
```diff
@@ -1,29 +0,0 @@
-
-description: ICP10125 barometric pressure/temperature sensor
-
-compatible: "invensense,icp10125"
```
Needs to be mentioned in the migration guide

yperess commented on drivers/sensor/tdk/icp101xx/icp101xx_drv.h:
```diff
@@ -0,0 +1,38 @@
+struct icp101xx_data {
+	int raw_pressure;
+	int raw_temperature;
+#ifdef ICP101XX_DRV_USE_FLOATS
+	float pressure;
+	float temperature;
+#else
+	int32_t pressure;
+	int32_t temperature;
+#endif
```
May I suggest to use the `q31_t` from `<zephyr/dsp/types.h>` here instead of splitting on floats?

rbuisson-invn replied:
If this is ok for you, could you please approve the PR?

yperess replied:
Yup, was only a wishful suggestion. I don't mind the float.

rbuisson-invn committed:
modules: hal_tdk: Add icp101xx support

icp101xx are barometric pressure and temperature sensors.
https://invensense.tdk.com/smartpressure

Signed-off-by: Remi Buisson <remi.buisson@tdk.com>

rbuisson-invn committed:
drivers: sensor: icp101xx: Add icp101xx support

Use official TDK Invensense driver for icp101xx sensor in tdk_hal module.

Signed-off-by: Remi Buisson <remi.buisson@tdk.com>

rbuisson-invn committed:
dts: bindings: sensor: Add invensense icp101xx properties

Replace icp10125.
tests: sensor: rename icp10125 to icp101xx

Signed-off-by: Remi Buisson <remi.buisson@tdk.com>

yperess commented on drivers/sensor/tdk/icp101xx/icp101xx_drv.c:
```diff
@@ -0,0 +1,259 @@
+void inv_icp101xx_sleep_us(int us)
+{
+	k_sleep(K_USEC(us));
+}
```
Ack
